[Go] Full source for customer loyalty workflow

afitz0 opened the pull request:
See the [go branch](https://github.com/afitz0/customer-loyalty-workflow/tree/go) for full commit history if curious.

Commit:
Stage go files for review

cretz commented on go/activities.go:
```diff
@@ -0,0 +1,14 @@
+
+import (
+	"context"
+	"go.temporal.io/sdk/activity"
```
Running `go fmt` (though it may still be separate for this case as `goimports`) should separate stdlib imports from non-stdlib imports by a blank line. Usually simply saving in your IDE does this.

afitz0 replied:
interestingly, goland didn't do this. I've run `goimports` now.

cretz commented on go/activities.go:
```diff
@@ -0,0 +1,14 @@
+
+type Activities struct{}
+
+func (a *Activities) SendEmail(ctx context.Context, body string) (err error) {
```
```suggestion
func (*Activities) SendEmail(ctx context.Context, body string) error {
```

No need to name variables you don't use

cretz commented on go/common/constants.go:
```diff
@@ -0,0 +1,24 @@
+
+const (
+	TaskQueue                = "CustomerLoyaltyTaskQueue"
+	CustomerWorkflowIdFormat = "customer-%v"
```
```suggestion
	CustomerWorkflowIDFormat = "customer-%v"
```

cretz commented on go/common/constants.go:
Having a "common" package is not usually best for Go programs. Go usually wants constants in files they are needed and helpers in files or packages they are needed. Also, don't over-optimize for making all string literals as constants.

cretz commented on go/common/helpers.go:
The Go way would be to put these three lines where they are needed and not have this helper

cretz commented on go/shared.go:
```diff
@@ -0,0 +1,21 @@
+		Points        int
+		AccountActive bool
+	}
+)
```
You don't have to use the `type (` approach btw. I never do. You can just declare top-level, e.g.:

```suggestion
type CustomerInfo struct {
	CustomerId    string
	LoyaltyPoints int
	StatusLevel   int
	Status        status.Status
	Name          string
	Guests        map[string]struct{}
	AccountActive bool
}

type GetStatusResponse struct {
	Tier          status.Tier
	Points        int
	AccountActive bool
}
```

cretz commented on go/shared.go:
Ideally the folder name is the package name (even if you have to go a level deeper)

cretz commented on go/shared.go:
```diff
@@ -0,0 +1,21 @@
+
+type (
+	CustomerInfo struct {
+		CustomerId    string
```
```suggestion
		CustomerID    string
```

Go traditionally capitalizes ID

cretz commented on go/starter/main.go:
```diff
@@ -0,0 +1,46 @@
+
+	wf "github.com/afitz0/customer-loyalty-workflow"
+	"github.com/afitz0/customer-loyalty-workflow/common"
+	"github.com/afitz0/customer-loyalty-workflow/zapadapter"
```
I don't see this package. Regardless, should be a simple set of code, doesn't need its own package. The Go way is to have fewer packages.

cretz commented on go/go.mod:
```diff
@@ -0,0 +1,37 @@
+module github.com/afitz0/customer-loyalty-workflow
```
Best practice is for the module path to be the GH path. So `github.com/afitz0/customer-loyalty-workflow/go` would be better practice.

cretz commented on go/starter/main.go:
```diff
@@ -0,0 +1,46 @@
+		StatusLevel:   0,
+		Name:          "Customer",
+		Guests:        map[string]struct{}{},
+		AccountActive: true,
```
```suggestion
		CustomerId:    "123",
		Name:          "Customer",
		AccountActive: true,
```

Redundant

cretz commented on go/starter/main.go:
```diff
@@ -0,0 +1,46 @@
+		AccountActive: true,
+	}
+	workflowOptions := client.StartWorkflowOptions{
+		ID:        fmt.Sprintf(common.CustomerWorkflowIdFormat, customer.CustomerId),
```
A helper in the workflow package that creates workflow ID from customer ID instead of exposing a printf format constant would be clearer IMO

cretz commented on go/shared.go:
```diff
@@ -0,0 +1,21 @@
+		CustomerId    string
+		LoyaltyPoints int
+		StatusLevel   int
+		Status        status.Status
```
Is this a serializable model? You can't serialize an interface usually (though you can count on it being nil I guess during serialization, but that's a bit weird to have a non-serializable exposed field an an exposed supposed-to-be-serializable struct)

afitz0 replied:
removed with refactoring of `StatusLevel` as proposed in another comment.

cretz commented on go/status/status.go:
No need for a separate package, just put this in the primary package

cretz commented on go/status/status.go:
```diff
@@ -0,0 +1,118 @@
+
+import "github.com/afitz0/customer-loyalty-workflow/common"
+
+var Levels = []Tier{
```
It's really confusing to have status interface implemented by level, a levels var as a collection of tiers, etc. The naming seems a bit off. How about just a simple `StatusLevel` struct that has these fields and the needed methods? Don't even need the status to be stateful if you just want to store the level or points elsewhere.

For example (just typed here in GH, so probably not exactly accurate):

```go
type StatusLevel struct {
  Ordinal int
  Name string
  MinimumPoints int
  GuestsAllowed int
}

func newStatusLevel(ord int, name string, minPoints int, guests int) *StatusLevel {
  return &StatusLevel{Ordinal: ord, Name: name, MinimumPoints: minPoints, GuestsAllowed: guests}
}

var StatusLevels = []*StatusLevel {
  newStatusLevel(0, "Member", 0, 0),
  newStatusLevel(1, "Bronze", 500, 1),
  newStatusLevel(2, "Silver", 1000, 2),
  newStatusLevel(3, "Gold", 2000, 5),
  newStatusLevel(4, "Platinum", 5000, 10),
}

func StatusLevelForPoints(points int) *StatusLevel {
  for i, level := range StatusLevels {
    if i > 0 && points < level.MinimumPoints {
      return StatusLevels[i-1]
    }
  }
  return StatusLevels[len(StatusLevels)-1]
}

// Returns nil if already at lowest
func (s *StatusLevel) Previous() *StatusLevel {
  if s.Ordinal > 0 {
    return StatusLevels[s.Ordinal-1]
  }
  return nil
}
```

That is probably enough assuming you don't need some mutable value. I would recommend just keeping the current mutable points count in the workflow or wherever.

afitz0 replied:
I've shamelessly stolen this implementation. :) Thank you!

cretz commented on go/status/status.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+type (
+	Status interface {
```
Why abstract an interface out for this?

afitz0 replied:
removed with refactoring of `StatusLevel` as proposed in another comment.

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 10 * time.Second,
+		RetryPolicy: &temporal.RetryPolicy{
+			InitialInterval: time.Second,
```
Not needed to be set, this is already the default. Also, if you can handle 100s instead of 60s, then the default retry policy is already good enough for you

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+	logger.Info("Validating customer info.")
+	validateCustomerInfo(&customer)
+
+	if info.ContinuedExecutionRunID == "" {
```
I think it'd be clearer if you used a workflow input field to make this decision and just set that field to false or whatever on continue as new

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+	// Block on everything. Continue-As-New on history length; size of activities in this workflow are small enough
+	// that we'll hit the length thresholds well before any size threshold.
+	for customer.AccountActive && info.GetCurrentHistoryLength() < common.EventsThreshold {
+		selector.Select(ctx)
```
You should also add a handler for `ctx.Done()` so the workflow can be cancelled

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+	child = workflow.ExecuteChildWorkflow(ctx, CustomerLoyaltyWorkflow, guest)
+	// wait for it to start before returning.
+	err = child.GetChildWorkflowExecution().Get(ctx, nil)
+	return err, child
```
Why return the child if you're abandoning it?

afitz0 replied:
refactored to a local activity as suggested

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+	return nil
+}
+
+func startGuestWorkflow(ctx workflow.Context, guest CustomerInfo) (err error, child workflow.ChildWorkflowFuture) {
```
Looking at your other code, you're basically wanting to signal-with-start a guest workflow with no parent attachment. I would consider a local activity to use the Temporal client to signal with start the guest workflow.

afitz0 replied:
done

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+		if _, ok := err.(*serviceerror.WorkflowExecutionAlreadyStarted); ok {
+			emailToSend = common.EmailGuestInvited
+			previousTier := customer.Status.PreviousTier()
+			err = guestWorkflow.SignalChildWorkflow(ctx, common.SignalEnsureMinimumStatus, previousTier).
```
I am a bit surprised the child is populated in this use case. But I would suggest not using this approach since even though it may _technically_ work, it may not be a child of this workflow. I would use `SignalExternalWorkflow` with the identifiers in the `WorkflowExecutionAlreadyStarted` error (if that is the error being returned).

Though signal-with-start as mentioned in previous code may be clearer.

afitz0 replied:
removed this bit as part of refactoring to an activity

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+		if _, ok := err.(*serviceerror.WorkflowExecutionAlreadyStarted); ok {
+			emailToSend = common.EmailGuestInvited
+			previousTier := customer.Status.PreviousTier()
+			err = guestWorkflow.SignalChildWorkflow(ctx, common.SignalEnsureMinimumStatus, previousTier).
```
Can you explain this logic in a comment? Specifically I'm not understanding why only an existing known guest can get promoted to tier-below-this-one but not a new guest.

afitz0 replied:
I believe the move to SignalWithStart addresses this -- i.e., it was a bug to have only existing guests get the promotion.

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+				logger.Info("Failed to signal 'already started' guest account; child workflow likely closed.")
+				emailToSend = common.EmailGuestCanceled
+			} else if err != nil {
+				return errors.Wrapf(err, "Could not signal guest/child workflow for guest Id '%v'.", guestId)
```
```suggestion
				return fmt.Errorf("could not signal guest/child workflow for guest Id '%v': %w", guestId, err)
```

This is the more common way to wrap errors (and Go doesn't want error phrases capitalized)

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+		emailBody := fmt.Sprintf(common.EmailPromoted, minStatus.Name)
+		err := workflow.ExecuteActivity(ctx, activities.SendEmail, emailBody).Get(ctx, nil)
+		if err != nil {
+			logger.Error("Error running SendEmail activity.", "Error", err)
```
You are inconsistent with whether you log errors before returning them

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+	return nil
+}
+
+func signalInviteGuest(ctx workflow.Context, c workflow.ReceiveChannel, customer *CustomerInfo) (err error) {
```
```suggestion
func signalInviteGuest(ctx workflow.Context, c workflow.ReceiveChannel, customer *CustomerInfo) error {
```

There is no value in naming these returns

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+		})
+
+	// signal handler for adding guest
+	selector.AddReceive(workflow.GetSignalChannel(ctx, common.SignalInviteGuest),
```
With the logic you have here, your next guest invite may be held up if, say, a previous invite guest email is just continually failing

afitz0 replied:
Setting a max attempts on these activities now.

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+
+func queryGetGuests(ctx workflow.Context, customer CustomerInfo) ([]string, error) {
+	logger := workflow.GetLogger(ctx)
+	guestIds := make([]string, len(customer.Guests))
```
I would recommend making `customer.Guests` a `[]string` and just looping for dedupe. That way you have a predictable output order for these kinds of calls (i.e. first guest is first returned). As it stands now, every call to this will have them in a different order, which is technically fine, but in workflows we like to promote predictability.

afitz0 replied:
done

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,270 @@
+
+func queryGetGuests(ctx workflow.Context, customer CustomerInfo) ([]string, error) {
+	logger := workflow.GetLogger(ctx)
+	guestIds := make([]string, len(customer.Guests))
+
+	i := 0
+	for k := range customer.Guests {
+		guestIds[i] = k
+		i++
+	}
```
```suggestion
	guestIDs := make([]string, 0, len(customer.Guests))
	for k := range customer.Guests {
		guestIDs = append(guestIDs, k)
	}
```

No need to manually keep track of index, just set capacity of slice and append

Commit:
goimports

Commit:
Round of addressing PR feedback. Major things:

* Id -> ID
* unname error returns
* remove unused named values

Commit:
errors.Wrap -> fmt.Errorf

Commit:
add /go to module name

Commit:
remove need for helpers.go

Commit:
move constants to workflow package, removing need for common package.

Commit:
minor refactoring

Commit:
Addressing PR feedback as follows:

* Add status.go at root with simpler implementation
* Refactor workflow to use new status impl
* Refactor CustomerInfo.Guests to use a flat slice instead of a map.
* Move starting guest workflow into a local activity.
* Refactor tests to match.

Commit:
goimports

Commit:
handle workflow cancellation requests

Commit:
move zapadapter out of its own package

Commit:
move all to ./loyalty/

Commit:
remove unused dep

cretz commented on go/activities.go:
```diff
@@ -0,0 +1,43 @@
+		workflowOptions, CustomerLoyaltyWorkflow, guest)
+	target := &serviceerror.WorkflowExecutionAlreadyStarted{}
+	if errors.As(err, &target) {
+		return &GuestAlreadyCanceledError{msg: "Guest account cannot be recreated from a closed status."}
```
FWIW, this is just the equivalent of `return temporal.NewApplicationError("Guest account cannot be recreated from a closed status.", "GuestAlreadyCanceledError")`. There is no benefit to the separate error struct since we have to serialize to Temporal error anyways.

EDIT: I see this is may be working just because it's a local activity, I'd have to check. See later comment.

cretz commented on go/starter/main.go:
```diff
@@ -0,0 +1,41 @@
+	"log"
+
+	"go.temporal.io/sdk/client"
+
```
Not the biggest fan of import separation like this, but no big deal (I personally prefer only one blank line to separate stdlib vs non-stlib, everything else together)

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+	EmailPromoted           = "Congratulations! You've been promoted to '%v' status!"
+	EmailDemoted            = "Unfortunately, you've lost enough points to bump you down to '%v' status. 😞"
+	EmailCancelAccount      = "Sorry to see you go!"
+)
```
Questionable benefit to moving these format strings away from where they are used (it's often hard for readers/devs to know the params a phrase may need if the phrase is somewhere else, much less keep them in sync). But if you must put them up here, might as well not expose them (i.e. do not capitalize).

afitz0 replied:
I've at least not exposed them now. I want to keep them as consts here as I use them in tests to check if the right emails are [attempted to be] sent at the right time(s).

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+
+func (e *GuestAlreadyCanceledError) Error() string { return e.msg }
+
+func CustomerLoyaltyWorkflow(ctx workflow.Context, customer CustomerInfo) (err error) {
```
```suggestion
func CustomerLoyaltyWorkflow(ctx workflow.Context, customer CustomerInfo) error {
```

No need to name the return type

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+
+	info := workflow.GetInfo(ctx)
+	selector := workflow.NewSelector(ctx)
+	activities := Activities{}
```
```suggestion
	var activities Activities
```

Another way to do the same thing, though don't have to change

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+	}
+
+	logger.Info("Loyalty workflow completed.", "Customer", customer, "WorkflowCanceled", workflowCanceled)
+	return nil
```
```suggestion
	return ctx.Err()
```

Or some other way to return cancellation error

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+
+// CustomerWorkflowID generates a Workflow ID based on the given customer ID.
+func CustomerWorkflowID(customerID string) string {
+	return fmt.Sprintf("customer-%v", customerID)
```
```suggestion
	return "customer- + customerID
```

Can just use basic concatenation in these simple cases, but your choice

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+	if customer.StatusLevel == nil {
+		customer.StatusLevel = StatusLevelForPoints(customer.LoyaltyPoints)
+	}
+	if len(customer.Guests) == 0 {
+		customer.Guests = make([]string, customer.StatusLevel.GuestsAllowed)
+	}
```
```suggestion
```

This is not needed. A `nil` slice is just like an empty one.

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+	}
+}
+
+func addGuestToCustomer(customer *CustomerInfo, guestID string) error {
+	guestExists := false
+	for _, g := range customer.Guests {
+		guestExists = g == guestID
+	}
+	// replace first empty slot, or append if none
+	if !guestExists {
+		added := false
+		for i, g := range customer.Guests {
+			if len(g) == 0 {
+				customer.Guests[i] = guestID
+				added = true
+				break
+			}
+		}
+		if !added {
+			customer.Guests = append(customer.Guests, guestID)
+		}
+	}
+
+	return nil
+}
```
```suggestion
func (c *Customer) addGuest(guestID string) {
	// Add if not there
	for _, g := range c.Guests {
		if g == guestID {
			return
		}
	}
	c.Guests = append(c.Guests, guestID)
}
```

This is the more Go-like way (once you remove the setting of the `Guests` to an empty pre-sized slice). Also, seems clearer to make this a method on the customer.

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+
+		err = workflow.ExecuteLocalActivity(ctx, activities.StartGuestWorkflow, guest).Get(ctx, nil)
+
+		t := &GuestAlreadyCanceledError{}
+		if errors.As(err, &t) {
```
Hrmm, I think this only works because it's a local activity. This wouldn't work in a normal activity (errors are not serialized back to their original type). I fear users will see this and think it's an ok thing to do w/ normal activity errors. I think we should either use application error and check the type, or not use error and use some return value instead.

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+	var guestID string
+	c.Receive(ctx, &guestID)
+
+	logger.Info("Checking to see if customer has enough status to allow for a guest invite.", "Customer", customer)
+	nonEmptyGuests := 0
+	for _, v := range customer.Guests {
+		if len(v) > 0 {
+			nonEmptyGuests++
+		}
+	}
```
Once you change the initialization of the guests slice, you can just use `len(customer.Guests)` to see how many guests

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+	var minStatus StatusLevel
+	c.Receive(ctx, &minStatus)
+
+	minOrd := minStatus.Ordinal
+	currentOrd := customer.StatusLevel.Ordinal
+
+	if currentOrd < minOrd {
```
```suggestion
	if customer.StatusLevel.Ordinal < minStatus.Ordinal {
```

Often it's just bloat to overly assign local variables for a single use

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+func signalEnsureMinimumStatus(ctx workflow.Context, c workflow.ReceiveChannel, customer *CustomerInfo) error {
+	activities := Activities{}
+
+	var minStatus StatusLevel
+	c.Receive(ctx, &minStatus)
```
I think you should send/accept something to represent the status (ordinal, points, etc) that you then use to look up the status with instead of having the caller send the entire status structure

cretz commented on go/workflow.go:
```diff
@@ -0,0 +1,333 @@
+	validateCustomerInfo(&customer)
+
+	if info.ContinuedExecutionRunID == "" {
+		err = workflow.ExecuteActivity(ctx, activities.SendEmail,
+			fmt.Sprintf(EmailWelcome, customer.StatusLevel.Name)).
+			Get(ctx, nil)
```
Should note on all of these emails, you are going to just hang the workflow forever if the email is failing to send. I don't see a problem with it, especially if email is a required part of this, just letting you know. Some people only try to send an email so many times (and increase the backoff between each from the default of rapid repeat) and don't let it stop down the workflow.

afitz0 replied:
👍 removed failing the workflow on SendEmail activity failure, and set a max attempts.

Commit:
New round of PR feedback:

* EnsureMinimumStatus signal takes ordinal rather than full signal.
* StartGuestWorkflow activity sends NewApplicationError instead of custom error type.
* Adding guest and validating customer info are now methods on CustomerInfo.
* Un-expose email string constants.
* Set max retry limit on activities.
* No longer err out workflow on SendEmail failure.
* Add flag to workflow inputs for `newCustomer`
* Misc small code formatting/style update.

cretz commented on go/loyalty/shared.go:
```diff
@@ -0,0 +1,32 @@
+type CustomerInfo struct {
+	CustomerID    string
+	LoyaltyPoints int
+	StatusLevel   *StatusLevel
```
Still a bit confusing to put this in the serialized data model when it can be derived from the points, but not a big deal

cretz commented on go/loyalty/activities.go:
```diff
@@ -0,0 +1,45 @@
+		return err
+	}
+
+	return nil
```
```suggestion
	}
	return err
```

(nitpick, can ignore)

cretz commented on go/loyalty/workflow.go:
```diff
@@ -0,0 +1,295 @@
+		return ctx.Err()
+	} else {
+		return nil
+	}
```
```suggestion
	}
	return nil
```

(nit, can ignore)

cretz commented on go/loyalty/workflow.go:
```diff
@@ -0,0 +1,295 @@
+	return "customer-" + customerID
+}
+
+func signalAddPoints(ctx workflow.Context, c workflow.ReceiveChannel, customer *CustomerInfo) error {
```
Why return an error from this call (or others where you never actually do return an error anymore)?

afitz0 replied:
removed the error return

cretz replied:
Hrmm, I can't unresolve things here. But it seems like the `error` is still there (maybe haven't pushed yet?)

afitz0 replied:
Sorry, yes, forgot to push. I have now.

cretz reviewed (approved):
Made a couple of minor comments, but looks good to me. May want others' opinions.

Commit:
Remove unused error returns and simplify a couple of if/else/return blocks.

Commit:
Remove StatusLevel field in CustomerInfo, replacing with lookup based on point value.

Commit:
Refactor StartGuestWorkflow:

- return enum rather than error since "guest canceled" is considered a case in normal operation, not error case.
- Also switch it to normal activity rather than local.

Commit:
Drain signals before CAN

Commit:
Receive on the signal channels before passing to named functions.
